nav2_system_tests/src/planning/planner_tester.cpp

```diff
@@ -59,22 +69,102 @@ PlannerTester::PlannerTester()
   startCostmapServer();
 
   // Launch a thread to process the messages for this node
-  spin_thread_ = new std::thread(&PlannerTester::spinThread, this);
+  spin_thread_ = std::make_unique<std::thread>(
+    [&]()
+    {
+      executor_.add_node(this->get_node_base_interface());
+      executor_.spin();
+      executor_.remove_node(this->get_node_base_interface());
+    });
 }
 
-PlannerTester::~PlannerTester()
+void PlannerTester::deactivate()
 {
+  if (!is_active_) {
+    RCLCPP_WARN(this->get_logger(), "Trying to deactivate while already inactive");
+    return;
+  }
+  is_active_ = false;
+
   executor_.cancel();
   spin_thread_->join();
-  delete spin_thread_;
-  spin_thread_ = nullptr;
+  spin_thread_.reset();
+
+  planner_client_.reset();
+  map_timer_.reset();
+  map_pub_.reset();
+  map_.reset();
+  costmap_server_.reset();
+  transform_publisher_.reset();
+}
+
+PlannerTester::~PlannerTester()
+{
+  if (is_active_) {
+    deactivate();
+  }
+}
+
+void PlannerTester::startPoseTransform()
+{
+  // Provide the robot pose transform
+  transform_publisher_ = create_publisher<tf2_msgs::msg::TFMessage>("/tf", rclcpp::QoS(100));
+
+  // Set an initial pose
+  geometry_msgs::msg::Point robot_position;
+  robot_position.x = 0.0;
+  robot_position.y = 0.0;
+  updateRobotPosition(robot_position);
+
+  // Publish the transform periodically
+  transform_timer_ = create_wall_timer(
```

Comment:
Can't you just publish this once when it changes? Just call publish at the end of updateRobotPosition?

How do you keep everything in sync if you don't do this? For instance - you start a plannerTest, set robot position to (x, y)  and send the request to compute path - the updated tf might not get published for 10ms

Reply:
If you just publish it once and the costmap fails to get it in time, then the entire test will fail since the timeout is also 1s. 

Example: https://2597-135363400-gh.circle-artifacts.com/0/opt/overlay_ws/log/test/nav2_system_tests/stdout.log

I'm trying this to fix a bug in that ticket as well.

Reply:
Typically when you have a server publishing some transformation you have a background server updating the transform so it doesnt time out. 
Ex.
https://github.com/SteveMacenski/slam_toolbox/blob/melodic-devel/src/slam_toolbox_common.cpp#L169

https://github.com/ros-perception/slam_karto/blob/melodic-devel/src/slam_karto.cpp#L317

https://github.com/cra-ros-pkg/robot_localization/blob/109dc784e9696bb487af11acdd7368c86863b362/src/ros_filter.cpp#L1813

Reply:
> If you just publish it once and the costmap fails to get it in time, then the entire test will fail since the timeout is also 1s.

My thinking is that it is reasonable to have the test fail in this case. I don't see any reason it should take so long to process the tf. If it takes a whole second, something is wrong.

However, I understand that this isn't a TF test and it may be necessary given the current state of affairs with the current TF implementation. I was hoping we could simplify given that this is just a test with a supposedly controlled environment.

Reply:
> Can't you just publish this once when it changes? Just call publish at the end of updateRobotPosition?

Agree, I forgot to add that. Pushed a change.

Reply:
@SteveMacenski @crdelsey Publishing the transform at the end of `updateRobotPosition` should be enough, however, I'm seeing that navfn sometimes doesn't receive the transform (that's why the tests are failing currently), hence the `tranform_timer_` -- maybe the initial pose transform is not received, but the following update might.